### ISSUE TITLE
fsharpi gui message

### ISSUE THREAD

**Issue by callmekohei** [fsharp/fsharp#805]
Please provide a succinct description of the issue.
```
fsharpi can not work well (^_^::
```

#### Repro steps

```
$ fsharpi

Microsoft (R) F# Interactive version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.

For help type #help;;

> Your system doesn't seem to support WinForms correctly. You will
need to set fsi.EventLoop use GUI windows from F# Interactive.
You can set different event loops for MonoMac, Gtk#, WinForms and other
UI toolkits. Drop the --gui argument if no event loop is required.
```

#### Expected behavior

```
$ fsharpi

F# Interactive for F# 4.1
Freely distributed under the Apache 2.0 Open Source License

For help type #help;;

> 
```

#### Actual behavior
```
$ fsharpi

Microsoft (R) F# Interactive version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.

For help type #help;;

> Your system doesn't seem to support WinForms correctly. You will
need to set fsi.EventLoop use GUI windows from F# Interactive.
You can set different event loops for MonoMac, Gtk#, WinForms and other
UI toolkits. Drop the --gui argument if no event loop is required.
```
#### Known workarounds
```
$ brew switch mono 5.0.1.1
Cleaning /usr/local/Cellar/mono/5.0.1.1
Cleaning /usr/local/Cellar/mono/5.4.1.6
227 links created for /usr/local/Cellar/mono/5.0.1.1
```

#### Related information 

* Operating system
```
$ sw_vers 
ProductName:	Mac OS X
ProductVersion:	10.13.2
BuildVersion:	17C88
```

* .NET Runtime, CoreCLR or Mono Version
```
$ mono --version
Mono JIT compiler version 5.4.1.6 (tarball Mon Dec 11 14:59:42 GMT 2017)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	LLVM:          supported, not enabled.
	GC:            sgen (concurrent by default)
```

**Comment by dsyme** [comment]
@callmekohei My understanding this has been resolved in the latest version of Mono.

@nosami Is that correct?  I'm concerned I don't know why this error message started appearing.  The GUI loop should not be installed by default when running on Mono:

```
    let mutable gui        = not runningOnMono // override via "--gui", on by default except when on Mono
```

**Comment by nosami** [comment]
I haven't seen that error before. I assume it's caused by bad compilation options in the homebrew formula. We don't support that.

```
14:51 $ mono --version
Mono JIT compiler version 5.8.0.103 (2017-10/1359c52fc6e Wed Dec 20 10:52:54 EST 2017)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug
	LLVM:          yes(3.6.0svn-mono-master/8b1520c8aae)
	GC:            sgen (concurrent by default)
✔ ~/src/FSharp.Compiler.Service/fcs [GetDeclarationListSymbols L|…9⚑ 4]
14:51 $ installpkg ~/Downloads/MonoFramework-MDK-5.4.0.
MonoFramework-MDK-5.4.0.130.macos10.xamarin.universal.pkg
MonoFramework-MDK-5.4.0.136.macos10.xamarin.universal.pkg
MonoFramework-MDK-5.4.0.147.macos10.xamarin.universal.pkg
MonoFramework-MDK-5.4.0.149.macos10.xamarin.universal.pkg
MonoFramework-MDK-5.4.0.193.macos10.xamarin.universal.pkg
MonoFramework-MDK-5.4.0.199.macos10.xamarin.universal.pkg
MonoFramework-MDK-5.4.0.201.macos10.xamarin.universal.pkg
MonoFramework-MDK-5.4.0.73.macos10.xamarin.universal.pkg
✔ ~/src/FSharp.Compiler.Service/fcs [GetDeclarationListSymbols L|…9⚑ 4]
14:51 $ installpkg ~/Downloads/MonoFramework-MDK-5.4.0.201.macos10.xamarin.universal.pkg
installer: Package name is Mono Framework
installer: Upgrading at base path /
installer: The upgrade was successful.
✔ ~/src/FSharp.Compiler.Service/fcs [GetDeclarationListSymbols L|…9⚑ 4]
14:53 $ fsharpi

Microsoft (R) F# Interactive version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.

For help type #help;;

> ✔ ~/src/FSharp.Compiler.Service/fcs [GetDeclarationListSymbols L|…9⚑ 4]
14:54 $ mono --version
Mono JIT compiler version 5.4.0.201 (2017-06/71277e78f6e Thu Sep 21 19:22:55 EDT 2017)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug
	LLVM:          yes(3.6.0svn-mono-master/8b1520c8aae)
	GC:            sgen (concurrent by default)
```

**Comment by dsyme** [comment]
If someone ants to upgrade the homebrew formula it is here: https://github.com/Homebrew/homebrew-core/blob/master/Formula/mono.rb and is currently at commit 35a4a5b1f26927259c3213465a47b27ffcd5cb4d https://github.com/Homebrew/homebrew-core/blob/master/Formula/mono.rb#L36

**Comment by baronfel** [comment]
There's been a few blockers around upgrading the version, mostly relating to how we don't have a standalone Homebrew distribution of mabuild 15 yet, which if I recall fsharp now requires to build, so the current upgrade PRs all got blocked on trying to include MSBuild the correct way. I think we should first investigate making an MSBuild for Homebrew, then just include that as a dependent package and then the fsharp upgrade should become more simple

**Comment by baronfel** [comment]
I take it back, looks like they've resolved the msbuild problem already! That's a happy surprise

**Comment by baronfel** [comment]
@nosami I have to retract my previous. The MSBuild PR was closed, not merged.  I'm going to spend the morning sorting through why and see if I can get some movement going. There seems to be a disconnect between the way Microsoft has packaged the dotnet sdk/cli and the way the homebrew project expects things to be.

**Comment by nosami** [comment]
We build fsharp using the Mono version of `msbuild` https://github.com/mono/msbuild/ . No idea if that is an option for you.

**Comment by baronfel** [comment]
Is this the build script you all are using? https://github.com/mono/mono/blob/master/packaging/MacSDK/msbuild.py

Just want to make sure I'm barking up the correct tree.

**Comment by nosami** [comment]
Yes, that's right if you want master. Branch 2017-10 is Mono 5.8, 2017-12 is is Mono 5.10 etc

**Comment by baronfel** [comment]
Which branch would be appropriate for 5.4? Thats the current stable and what the mono homebrew package is currently using.

**Comment by nosami** [comment]
2017-06

**Comment by callmekohei** [comment]
Hello! I'm callmekohei!

So, thank you for your advieces!

I'm glad (^_^)/

Now I will pull request to `Home brew`.

First of all, I think to do following.
```
01. install mono    <--- I can do it!
02. install msbuild <--- I can not do it!
03. install fsharp  <--- I can not do it!
```

I'm glad of you to give me some hints.

Thank you.

---

### I do following...

01. install mono 
```bash
$ vim foo.bash

#!/bin/bash

# Replace PREFIX with your installation prefix(path) that you selected:
PREFIX=/Users/callmekohei/mono64

# make folder for mono
mkdir /Users/callmekohei/mono64

# download source of mono
git clone --depth 1 https://github.com/mono/mono.git

# move mono folder
cd mono

# build and install mono
./autogen.sh --prefix=$PREFIX --disable-nls
make
sudo make install


$ bash foo.bash


--- wait a long long time... ---


$ mono --version

Mono JIT compiler version 5.11.0 (master/0d9c3dff 2018年 1月16日 火曜日 14時04分58秒 JST)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          supported, not enabled.
	GC:            sgen (concurrent by default)

```

02. install msbuild 

```bash
// download msbuild source
$ git clone --depth 1 https://github.com/Microsoft/msbuild


// link openssl
$ ls -sf /usr/local/Cellar/openssl/1.0.2m/lib/libcrypto.1.0.0.dylib /usr/local/lib
$ ls -sf /usr/local/Cellar/openssl/1.0.2m/lib/libssl.1.0.0.dylib /usr/local/lib


// build msbuild
$ ./cibuild.sh --scope Compile --target Mono --host Mono

debug: which mono: /Users/callmekohei/mono64/bin/mono
MONO_BIN_DIR: 
** Using mono from /Users/callmekohei/mono64/bin/mono
Mono JIT compiler version 5.11.0 (master/0d9c3dff 2018年 1月16日 火曜日 14時04分58秒 JST)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          supported, not enabled.
	GC:            sgen (concurrent by default)
Installing dotnet cli...
Restoring BuildTools version 1.0.27-prerelease-00927-05...
/Users/callmekohei/tmp/msbuild/init-tools.sh: line 132: 47300 Segmentation fault: 11  $__DOTNET_CMD restore "$__PROJECT_JSON_FILE" --no-cache --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE $__INIT_TOOLS_RESTORE_ARGS >> $__init_tools_log
ERROR: Could not restore build tools correctly. See '/Users/callmekohei/tmp/msbuild/init-tools.log' for more details.1
Initializing BuildTools...
/Users/callmekohei/tmp/msbuild/init-tools.sh: line 122: /Users/callmekohei/tmp/msbuild/packages/Microsoft.DotNet.BuildTools/1.0.27-prerelease-00927-05/lib/init-tools.sh: No such file or directory
ERROR: An error occured when trying to initialize the tools. Please check '/Users/callmekohei/tmp/msbuild/init-tools.log' for more details.1
```


03. install fsharp 

```
// create msbuild command
$ vim msbuild

/Users/callmekohei/tmp/msbuild/packages/msbuild/msbuild $@


// change mod
$ chmod 777 msbuild


// regist command to OSX
$ mv msbuild /usr/local/bin/


// check
$ msbuild /version

Microsoft (R) Build Engine version 14.1.0.0
Copyright (C) Microsoft Corporation. All rights reserved.

14.1.0.0

// download source of fsharp
$ git clone --depth 1 https://github.com/fsharp/fsharp


// move fsharp folder
$ cd fsharp/


// set pkg config path
$ export PKG_CONFIG_PATH=/Users/callmekohei/mono64/lib/pkgconfig


// autogen
$ ./autogen.sh --prefix /Users/callmekohei/mono64

checking whether make sets $(MAKE)... yes
checking for pkg-config... /usr/local/bin/pkg-config
configure: "pkg-config: /usr/local/bin/pkg-config"
configure: "PKG_CONFIG_LIBDIR: "
checking for msbuild... /usr/local/bin/msbuild
configure: msbuild: /usr/local/bin/msbuild
configure: creating ./config.status
config.status: creating mono/launcher
config.status: creating mono/config.make


// make
$ make

/Applications/Xcode.app/Contents/Developer/usr/bin/make build-proto
MONO_ENV_OPTIONS= mono .nuget/NuGet.exe restore packages.config -PackagesDirectory packages -ConfigFile .nuget/NuGet.Config
All packages listed in packages.config are already installed.
MONO_ENV_OPTIONS= /usr/local/bin/msbuild /p:Configuration=Proto /p:TargetDotnetProfile=net40 src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj
Microsoft (R) Build Engine version 14.1.0.0
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 1/16/2018 8:32:23 PM.
Project "/Users/callmekohei/tmp/fsharp/src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj" on node 1 (default targets).
CallFsSrGen:
Skipping target "CallFsSrGen" because all output files are up-to-date with respect to the input files.
The target "Localize" listed in an AfterTargets attribute at "/Users/callmekohei/tmp/fsharp/src/fsharp/FSharp.Core/FSharp.Core.fsproj (284,43)" does not exist in the project, and will be ignored.
Project "/Users/callmekohei/tmp/fsharp/src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj" (1) is building "/Users/callmekohei/tmp/fsharp/src/fsharp/FSharp.Core/FSharp.Core.fsproj" (2) on node 1 (default targets).
/Users/callmekohei/tmp/fsharp/packages/FSharp.Compiler.Tools.4.1.27/tools/Microsoft.FSharp.Targets(135,13): error MSB4131: The "ResourceFilesWithManifestResourceNames" parameter is not supported by the "CreateFSharpManifestResourceName" task. Verify the parameter exists on the task, and it is a gettable public instance property. [/Users/callmekohei/tmp/fsharp/src/fsharp/FSharp.Core/FSharp.Core.fsproj]
Done Building Project "/Users/callmekohei/tmp/fsharp/src/fsharp/FSharp.Core/FSharp.Core.fsproj" (default targets) -- FAILED.
Done Building Project "/Users/callmekohei/tmp/fsharp/src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj" (default targets) -- FAILED.

Build FAILED.

"/Users/callmekohei/tmp/fsharp/src/fsharp/FSharp.Build-proto/FSharp.Build-proto.fsproj" (default target) (1) ->
"/Users/callmekohei/tmp/fsharp/src/fsharp/FSharp.Core/FSharp.Core.fsproj" (default target) (2) ->
(CreateManifestResourceNames target) -> 
  /Users/callmekohei/tmp/fsharp/packages/FSharp.Compiler.Tools.4.1.27/tools/Microsoft.FSharp.Targets(135,13): error MSB4131: The "ResourceFilesWithManifestResourceNames" parameter is not supported by the "CreateFSharpManifestResourceName" task. Verify the parameter exists on the task, and it is a gettable public instance property. [/Users/callmekohei/tmp/fsharp/src/fsharp/FSharp.Core/FSharp.Core.fsproj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.90
make[1]: *** [build-proto] Error 1
make: *** [all] Error 2
```

**Comment by nosami** [comment]
Hi @callmekohei! msbuild didn't install correctly. You are still using 14.1.0.0 but you should be using 15+

Did you try to install from https://github.com/mono/msbuild/ _not_ https://github.com/Microsoft/msbuild ?

**Comment by callmekohei** [comment]
Hello! @nosami! 

Yes! (^_^)/
```
// download msbuild source
$ git clone --depth 1 https://github.com/Microsoft/msbuild
```

**Comment by callmekohei** [comment]
I do not know about mono version 14...

why (^_^;;;

```
$ msbuild /version

Microsoft (R) Build Engine version 14.1.0.0
Copyright (C) Microsoft Corporation. All rights reserved.

14.1.0.0
```

**Comment by nosami** [comment]
That is the wrong one to use :) Try https://github.com/mono/msbuild/ instead. 


If you don't need to use HomeBrew, then I'd just recommend using the `.pkg` instead (it has Mono, msbuild & fsharp ready to use)... but you would need to unlink the homebrew packages

**Comment by nosami** [comment]
```
11:56 $ msbuild /version
Microsoft (R) Build Engine version 15.4.0.0 (xplat-master/5cf9eb2c Wed Dec 20 11:07:34 EST 2017) for Mono
Copyright (C) Microsoft Corporation. All rights reserved.
```

This is what I am using right now.

**Comment by callmekohei** [comment]
I just re-try now! (^_^)/

**Comment by callmekohei** [comment]
Hmmm....

**Comment by callmekohei** [comment]
Just error (^_^;;;

```shell
tmp [master]$ git clone  https://github.com/mono/msbuild/
Cloning into 'msbuild'...
remote: Counting objects: 32414, done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 32414 (delta 2), reused 1 (delta 1), pack-reused 32408
Receiving objects: 100% (32414/32414), 39.48 MiB | 638.00 KiB/s, done.
Resolving deltas: 100% (25483/25483), done.
tmp [master]$ cd msbuild/
msbuild [xplat-master]$ ./cibuild.sh --scope Compile --target Mono --host Mono
debug: which mono: /Users/callmekohei/mono64/bin/mono
MONO_BIN_DIR: 
** Using mono from /Users/callmekohei/mono64/bin/mono
Mono JIT compiler version 5.11.0 (master/0d9c3dff 2018年 1月16日 火曜日 14時04分58秒 JST)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          supported, not enabled.
	GC:            sgen (concurrent by default)
** Downloading MSBUILD from https://github.com/Microsoft/msbuild/releases/download/mono-hosted-msbuild-v0.03/mono_msbuild_d25dd923839404bd64cc63f420e75acf96fc75c4.zip
Installing dotnet cli...
Restoring BuildTools version 1.0.27-prerelease-00927-05...
/Users/callmekohei/tmp/msbuild/init-tools.sh: line 137: 48087 Segmentation fault: 11  $__DOTNET_CMD restore "$__PROJECT_JSON_FILE" --no-cache --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE $__INIT_TOOLS_RESTORE_ARGS >> $__init_tools_log
ERROR: Could not restore build tools correctly. See '/Users/callmekohei/tmp/msbuild/init-tools.log' for more details.
============= /Users/callmekohei/tmp/msbuild/init-tools.log ===========
Running: /Users/callmekohei/tmp/msbuild/init-tools.sh
Installing 'https://dotnetcli.blob.core.windows.net/dotnet/Sdk/1.0.0-preview3-003686/dotnet-dev-osx-x64.1.0.0-preview3-003686.tar.gz' to '/Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet.tar'
Running: /Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet restore "/Users/callmekohei/tmp/msbuild/Tools/1.0.27-prerelease-00927-05/project.json" --no-cache --packages /Users/callmekohei/tmp/msbuild/packages --source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json --legacy-packages-directory
============= end of /Users/callmekohei/tmp/msbuild/init-tools.log ===========
Initializing BuildTools...
/Users/callmekohei/tmp/msbuild/init-tools.sh: line 127: /Users/callmekohei/tmp/msbuild/packages/Microsoft.DotNet.BuildTools/1.0.27-prerelease-00927-05/lib/init-tools.sh: No such file or directory
ERROR: An error occured when trying to initialize the tools. Please check '/Users/callmekohei/tmp/msbuild/init-tools.log' for more details.1
msbuild [xplat-master]$ 
```

**Comment by nosami** [comment]
what if you `git co 5cf9eb2c`?

**Comment by callmekohei** [comment]
？？
`git co 5cf9eb2c`

git version?
```
tmp [master]$ git --version
git version 2.14.3 (Apple Git-98)
hub version 2.2.9
```

**Comment by callmekohei** [comment]
```
msbuild [xplat-master]$ gittree | head
*   	 9a983c1e Mon Jan 8 12:29:14 2018 -0500 Merge pull request #34 from radical/xplat-master-update  (HEAD -> xplat-master, origin/xplat-master, origin/HEAD)
|\  
| * 	 4d05a368 Thu Dec 21 15:32:50 2017 -0500 [tests] Enable and fix tests that failed on mono-osx 
| * 	 25b09b14 Thu Dec 21 15:15:54 2017 -0500 [tests] Fix hardcoded windows paths to be more xplat 
| * 	 49c624e6 Thu Dec 21 15:13:53 2017 -0500 [tests] Fix hardcoded windows paths to be more xplat 
| * 	 45698265 Thu Dec 21 15:13:09 2017 -0500 [tests] Fix hardcoded windows paths to be more xplat 
| * 	 cdcd09ce Wed Dec 20 16:34:43 2017 -0500 [tests] Fix hardcoded windows paths to be more xplat 
| * 	 09d550b1 Wed Dec 20 16:15:55 2017 -0500 [mono][tests] Disable evaluation profiling tests on mono as they depend 
| *   	 6f740e45 Wed Dec 20 16:04:37 2017 -0500 Merge remote-tracking branch 'origin/master' into xplat-master 
```

**Comment by nosami** [comment]
This https://github.com/mono/mono/blob/master/packaging/MacSDK/msbuild.py is the 'recipe' that we use to build msbuild.

Mono master is currently at commit `abbd02a3b0fa710c5a208b62bc0ba944b0a5badf`

**Comment by callmekohei** [comment]
Oh! this!
```
msbuild [xplat-master]$ gittree | grep 'abbd02a'
* | 	 abbd02a3 Tue Nov 28 17:04:45 2017 +0100 Explicitly copy files from inside ExtensionsPath to xbuild dirs (#33) 
```

**Comment by callmekohei** [comment]
I just re-try! (^_^)/

**Comment by callmekohei** [comment]
```
msbuild [xplat-master]$ git reset --hard abbd02a3
HEAD is now at abbd02a3 Explicitly copy files from inside ExtensionsPath to xbuild dirs (#33)
```

**Comment by callmekohei** [comment]
Hmmm....
```
msbuild [xplat-master]$ ./cibuild.sh --scope Compile --target Mono --host Mono
debug: which mono: /Users/callmekohei/mono64/bin/mono
MONO_BIN_DIR: 
** Using mono from /Users/callmekohei/mono64/bin/mono
Mono JIT compiler version 5.11.0 (master/0d9c3dff 2018年 1月16日 火曜日 14時04分58秒 JST)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          supported, not enabled.
	GC:            sgen (concurrent by default)
Installing dotnet cli...
Restoring BuildTools version 1.0.27-prerelease-00927-05...
/Users/callmekohei/tmp/msbuild/init-tools.sh: line 137: 48206 Segmentation fault: 11  $__DOTNET_CMD restore "$__PROJECT_JSON_FILE" --no-cache --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE $__INIT_TOOLS_RESTORE_ARGS >> $__init_tools_log
ERROR: Could not restore build tools correctly. See '/Users/callmekohei/tmp/msbuild/init-tools.log' for more details.
============= /Users/callmekohei/tmp/msbuild/init-tools.log ===========
Running: /Users/callmekohei/tmp/msbuild/init-tools.sh
Installing 'https://dotnetcli.blob.core.windows.net/dotnet/Sdk/1.0.0-preview3-003686/dotnet-dev-osx-x64.1.0.0-preview3-003686.tar.gz' to '/Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet.tar'
Running: /Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet restore "/Users/callmekohei/tmp/msbuild/Tools/1.0.27-prerelease-00927-05/project.json" --no-cache --packages /Users/callmekohei/tmp/msbuild/packages --source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json --legacy-packages-directory
log  : Restoring packages for /Users/callmekohei/tmp/msbuild/Tools/1.0.27-prerelease-00927-05/project.json...
============= end of /Users/callmekohei/tmp/msbuild/init-tools.log ===========
Initializing BuildTools...
/Users/callmekohei/tmp/msbuild/init-tools.sh: line 127: /Users/callmekohei/tmp/msbuild/packages/Microsoft.DotNet.BuildTools/1.0.27-prerelease-00927-05/lib/init-tools.sh: No such file or directory
ERROR: An error occured when trying to initialize the tools. Please check '/Users/callmekohei/tmp/msbuild/init-tools.log' for more details.1
```

**Comment by callmekohei** [comment]
why...

**Comment by nosami** [comment]
I'm not sure... I just tried and got the same result. Maybe @radical can help?

**Comment by callmekohei** [comment]
### log file
```
Running: /Users/callmekohei/tmp/msbuild/init-tools.sh
Installing 'https://dotnetcli.blob.core.windows.net/dotnet/Sdk/1.0.0-preview3-003686/dotnet-dev-osx-x64.1.0.0-preview3-003686.tar.gz' to '/Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet.tar'
Running: /Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet restore "/Users/callmekohei/tmp/msbuild/Tools/1.0.27-prerelease-00927-05/project.json" --no-cache --packages /Users/callmekohei/tmp/msbuild/packages --source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json --legacy-packages-directory
log  : Restoring packages for /Users/callmekohei/tmp/msbuild/Tools/1.0.27-prerelease-00927-05/project.json...
Running: /Users/callmekohei/tmp/msbuild/packages/Microsoft.DotNet.BuildTools/1.0.27-prerelease-00927-05/lib/init-tools.sh /Users/callmekohei/tmp/msbuild /Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet /Users/callmekohei/tmp/msbuild/Tools
```

**Comment by callmekohei** [comment]
Thank you @nosami 

I'm glad of your help (^_^)/

**Comment by callmekohei** [comment]
data from 

https://github.com/Microsoft/msbuild/releases/tag/mono-hosted-msbuild-v0.03

**Comment by callmekohei** [comment]
just casue place...

```bash
a=/Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet
b=restore
c="/Users/callmekohei/tmp/msbuild/Tools/1.0.27-prerelease-00927-05/project.json"
d=--no-cache
e=--packages
f=/Users/callmekohei/tmp/msbuild/packages
g=--source
h=https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json
i=--legacy-packages-directory

$a $b $c $d $e $f $g $h $i
```
result
```
Restoring packages for
/Users/callmekohei/tmp/msbuild/Tools/1.0.27-prerelease-00927-05/project.json...
/Users/callmekohei/tmp/bar.bash:
line 11: 50641 Segmentation fault: 11
$a $b $c $d $e $f $g $h $i
```

**Comment by callmekohei** [comment]
$ cat /Users/callmekohei/tmp/msbuild/Tools/1.0.27-prerelease-00927-05/project.json

```json
{ "dependencies": { "Microsoft.DotNet.BuildTools": "1.0.27-prerelease-00927-05" }, "frameworks": { "netcoreapp1.0": 
```

**Comment by radical** [comment]
You are probably missing openssl bits. See https://github.com/Microsoft/msbuild/wiki/Building-Testing-and-Debugging-on-Mono-MSBuild

**Comment by callmekohei** [comment]
Hmmm....
```
/// https://github.com/Microsoft/msbuild/wiki/Building-Testing-and-Debugging-on-Mono-MSBuild#unix
The CLI requires a recent OpenSSL library available in /usr/lib
```

`/usr/lib` ??? not `/usr/local/lib`....

```
$ ls -al /usr/local/lib/ | grep 'ssl'
lrwxr-xr-x    1 callmekohei  admin    58  1 16 21:36 libcrypto.1.0.0.dylib -> /usr/local/Cellar/openssl/1.0.2m/lib/libcrypto.1.0.0.dylib
lrwxr-xr-x    1 callmekohei  admin    55  1 16 21:36 libssl.1.0.0.dylib -> /usr/local/Cellar/openssl/1.0.2m/lib/libssl.1.0.0.dylib
```

I do following
```
$ sudo ln -sf /usr/local/Cellar/openssl/1.0.2m/lib/libcrypto.1.0.0.dylib /usr/lib
Password:
ln: /usr/lib/libcrypto.1.0.0.dylib: Operation not permitted
$ sudo ln -sf /usr/local/Cellar/openssl/1.0.2m/lib/libssl.1.0.0.dylib /usr/lib
ln: /usr/lib/libssl.1.0.0.dylib: Operation not permitted
```

**Comment by callmekohei** [comment]
Oh...

paphaps..

OSX sierra vs dotnet 1.0.1
```
OSX sierra   ---> prohibit to access /usr/lib (System-Only Locations)

dotnet 1.0.1  ---> try access /usr/lib
```

see : https://developer.apple.com/library/content/documentation/Security/Conceptual/System_Integrity_Protection_Guide/FileSystemProtections/FileSystemProtections.html#//apple_ref/doc/uid/TP40016462-CH2-SW1

see : https://github.com/Microsoft/msbuild/wiki/Building-Testing-and-Debugging-on-.Net-Core-MSBuild#unix

**Comment by callmekohei** [comment]
Hello.

It's very difficult for me ... (^_^;;;

So I have done so far.

```
msbuild [xplat]$ ./cibuild.sh --scope Compile --target Mono --host Mono
debug: which mono: /Users/callmekohei/mono64/bin/mono
MONO_BIN_DIR: 
** Using mono from /Users/callmekohei/mono64/bin/mono
Mono JIT compiler version 5.11.0 (master/0d9c3dff 2018年 1月16日 火曜日 14時04分58秒 JST)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          supported, not enabled.
	GC:            sgen (concurrent by default)
Installing dotnet cli...
Initializing BuildTools...
Done initializing tools.

** Rebuilding MSBuild with downloaded binaries

** using MSBuild in path: /Users/callmekohei/tmp/msbuild/packages/msbuild/MSBuild.exe
** using runtime host in path: /Users/callmekohei/mono64/bin/mono
** /Users/callmekohei/mono64/bin/mono --debug /Users/callmekohei/tmp/msbuild/packages/msbuild/MSBuild.exe /t:Rebuild "/Users/callmekohei/tmp/msbuild/build.proj" /p:OS=OSX /p:Configuration=Debug-MONO /p:OverrideToolHost=/Users/callmekohei/mono64/bin/mono /verbosity:minimal  /p:CscToolExe=csc.exe /p:CscToolPath=/Users/callmekohei/tmp/msbuild/packages/msbuild/ /p:DebugType=portable /fl "/flp:v=diag;logfile=/Users/callmekohei/tmp/msbuild/msbuild_bootstrap_build-Mono.log"
Microsoft (R) Build Engine version 14.1.0.0
Copyright (C) Microsoft Corporation. All rights reserved.

  Restoring build tools...
  /var/folders/b1/23d5mpmn4mn1rh3b_7pf7wfw0000gn/T/tmp65c734712aaa428682b14b9cd2b0c9de.exec.cmd: line 3: 66248 Segmentation fault: 11  "/Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet" restore --legacy-packages-directory --packages /Users/callmekohei/tmp/msbuild/packages/ "/Users/callmekohei/tmp/msbuild/src/.nuget/project.json"
  log  : Restoring packages for /Users/callmekohei/tmp/msbuild/src/.nuget/project.json...
/Users/callmekohei/tmp/msbuild/src/.nuget/packageLoad.targets(41,5): error MSB3073: The command "ulimit -n 2500 ; "/Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet" restore --legacy-packages-directory --packages /Users/callmekohei/tmp/msbuild/packages/ "/Users/callmekohei/tmp/msbuild/src/.nuget/project.json"" exited with code 139. [/Users/callmekohei/tmp/msbuild/build.proj]
```

```
msbuild [xplat]$ /Users/callmekohei/mono64/bin/mono --debug /Users/callmekohei/tmp/msbuild/packages/msbuild/MSBuild.exe /t:Rebuild "/Users/callmekohei/tmp/msbuild/build.proj" /p:OS=OSX /p:Configuration=Debug-MONO /p:OverrideToolHost=/Users/callmekohei/mono64/bin/mono /verbosity:minimal  /p:CscToolExe=csc.exe /p:CscToolPath=/Users/callmekohei/tmp/msbuild/packages/msbuild/ /p:DebugType=portable /fl "/flp:v=diag;logfile=/Users/callmekohei/tmp/msbuild/msbuild_bootstrap_build-Mono.log"
Microsoft (R) Build Engine version 14.1.0.0
Copyright (C) Microsoft Corporation. All rights reserved.

  Building Project(s): build/dirs.proj : Configuration=Debug-MONO, Platform=AnyCPU 
  Building Project(s): src/dirs.proj : Configuration=Debug-MONO, Platform=AnyCPU 
  Building Project(s): Samples/dirs.proj : Configuration=Debug-MONO, Platform=AnyCPU 
/Users/callmekohei/tmp/msbuild/src/dir.targets(79,3): error MSB4019: The imported project "/Users/callmekohei/tmp/msbuild/packages/Nerdbank.GitVersioning/1.5.46/build/dotnet/Nerdbank.GitVersioning.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [/Users/callmekohei/tmp/msbuild/src/dirs.proj]
  /var/folders/b1/23d5mpmn4mn1rh3b_7pf7wfw0000gn/T/tmp2a91ce3ee64d4c15b2bce8795a69eeb1.exec.cmd: line 3: 66524 Segmentation fault: 11  "/Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet" restore --legacy-packages-directory --packages /Users/callmekohei/.nuget/packages "/Users/callmekohei/tmp/msbuild/Samples/TaskUsageLogger/project.json"
/Users/callmekohei/tmp/msbuild/Tools/packageresolve.targets(35,5): error MSB3073: The command "ulimit -n 2500 ; "/Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet" restore --legacy-packages-directory --packages /Users/callmekohei/.nuget/packages "/Users/callmekohei/tmp/msbuild/Samples/TaskUsageLogger/project.json"" exited with code 139. [/Users/callmekohei/tmp/msbuild/Samples/TaskUsageLogger/TaskUsageLogger.csproj]
/Users/callmekohei/tmp/msbuild/src/dir.targets(79,3): error MSB4019: The imported project "/Users/callmekohei/tmp/msbuild/packages/Nerdbank.GitVersioning/1.5.46/build/dotnet/Nerdbank.GitVersioning.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [/Users/callmekohei/tmp/msbuild/src/Framework/Microsoft.Build.Framework.csproj]
/Users/callmekohei/tmp/msbuild/src/dir.targets(79,3): error MSB4019: The imported project "/Users/callmekohei/tmp/msbuild/packages/Nerdbank.GitVersioning/1.5.46/build/dotnet/Nerdbank.GitVersioning.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [/Users/callmekohei/tmp/msbuild/src/Utilities/Microsoft.Build.Utilities.csproj]
  /var/folders/b1/23d5mpmn4mn1rh3b_7pf7wfw0000gn/T/tmpe8aee61c46ba42f1970fa911a7ccd976.exec.cmd: line 3: 66527 Segmentation fault: 11  "/Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet" restore --legacy-packages-directory --packages /Users/callmekohei/.nuget/packages "/Users/callmekohei/tmp/msbuild/Samples/Dependency/project.json"
/Users/callmekohei/tmp/msbuild/Tools/packageresolve.targets(35,5): error MSB3073: The command "ulimit -n 2500 ; "/Users/callmekohei/tmp/msbuild/Tools/dotnetcli/dotnet" restore --legacy-packages-directory --packages /Users/callmekohei/.nuget/packages "/Users/callmekohei/tmp/msbuild/Samples/Dependency/project.json"" exited with code 139. [/Users/callmekohei/tmp/msbuild/Samples/Dependency/Dependency.csproj]
/Users/callmekohei/tmp/msbuild/src/dir.targets(79,3): error MSB4019: The imported project "/Users/callmekohei/tmp/msbuild/packages/Nerdbank.GitVersioning/1.5.46/build/dotnet/Nerdbank.GitVersioning.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [/Users/callmekohei/tmp/msbuild/src/Framework/Microsoft.Build.Framework.csproj]
/Users/callmekohei/tmp/msbuild/src/dir.targets(79,3): error MSB4019: The imported project "/Users/callmekohei/tmp/msbuild/packages/Nerdbank.GitVersioning/1.5.46/build/dotnet/Nerdbank.GitVersioning.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [/Users/callmekohei/tmp/msbuild/src/Utilities/Microsoft.Build.Utilities.csproj]
```

**Comment by radical** [comment]
Use the `xplat-master` branch.

**Comment by callmekohei** [comment]
@radical 

Thank you!

I'll just try it (^_^)/

**Comment by baronfel** [comment]
I've had some trouble going through and working on the packaging in the way the homebrew team will take it:

- as an optional add-in (by default) to the existing mono install, the same way fsharp is currently, or
- as a dedicated homebrew formula for msbuild, which could then be depended on by a dedicated fsharp formula (see WIP here: https://github.com/baronfel/homebrew-core/blob/msbuild/Formula/msbuild.rb)

The homebrew team doesn't like the first formulation (see discussion here: https://github.com/Homebrew/homebrew-core/pull/21607), so I tried to do the second.  This way doesn't work because the msbuild install tries to overwrite files that are not writable (namely adding and placing tasks, targets, props inside the mono install). If we could somehow make the install add these files to a new location and add that location to the search path, that could suffice.

Given the amount of pain, is it more worth our while to stop advertising and deprecate the fsharp deployment in the homebrew mono formula, because it will just get more and more behind as long as msbuild is not able to be installed, and just advertise the existing homebrew cask install of the mono-mdk? In this was users will also get things like libgdiplus that currently are not part of the homebrew fsharp packaging.

**Comment by radical** [comment]
msbuild should be bundled with mono, installed in the same prefix. It assumes that is it installed in a specific location in the mono prefix. Eventually, we will probably build this from the mono build itself.

If the issue is openssl or building on High Sierra, then that is because msbuild build still depends on dotnet 1.x . There is ongoing work to port this to 2.0 which doesn't have openssl issues and works on High Sierra (https://github.com/Microsoft/msbuild/pull/2868).

**Comment by callmekohei** [comment]
I would like msbuild bundled with mono. (^_^)/

**Comment by callmekohei** [comment]
Hello! Hello! Hello! I'm `callmekohei`!

## Hmmm...

It's very very very tough.

I can not build `msbuild`.

I wait for @baronfel and [Microsoft/msbuild#2868](https://github.com/Microsoft/msbuild/pull/2868)

Thank you.

---


## I do following this ...

see also : [Mono MSBuild](https://github.com/Microsoft/msbuild/wiki/Building-Testing-and-Debugging-on-Mono-MSBuild)

see also : [init-tools.sh ](https://gist.github.com/callmekohei/b0e2e198ec382ef717e85a15616f0061)

see alse : [init-tools.log](https://gist.github.com/callmekohei/f7321910a8241b204987c990bc54a2bf)

```
// download mono msbuild source
$ git clone --depth 1 https://github.com/mono/msbuild/

// move msbuild folder
$ mv msbuild/

// re-write init-tools.sh
msbuild [xplat-master]$ vim init-tools.sh
```
`init-tools.sh`
```sh
#!/usr/bin/env bash

# ---------------------------------------------------------------------
# Util functions
# ---------------------------------------------------------------------

partingLine () {
  printf -v _hr "%*s" $(tput cols) && echo ${_hr// /${1--}}
}


# ---------------------------------------------------------------------
# Global variables
# ---------------------------------------------------------------------

# path and version

#   msbuild/
#   │
#   ├── init-tools.sh                        <---- __scriptpath
#   ├── Tools                                <---- __TOOLRUNTIME_DIR
#   │   ├── tool-runtime                     <---- __TOOLRUNTIME_DIR ( init-Tools.sh in packages/.../lib )
#   │   ├── 1.0.27-prerelease-00927-05/      <---- __BUILD_TOOLS_PATH
#   │   │   ├── done                         <---- __INIT_TOOLS_DONE_MARKER
#   │   │   └── project.csproj               <---- __PROJECT_JSON_PATH
#   │   └── dotnetcli                        <---- __DOTNET_PATH
#   ├── packages                             <---- __PACKAGES_DIR
#   │   └── microsoft.dotnet.buildtools
#           └── 1.0.27-prerelease-00927-05
#               └── lib                      <---- __BUILD_TOOLS_PATH
#                   └── init-tools.sh

__scriptpath=$(cd "$(dirname "$0")"; pwd -P)
__BUILD_TOOLS_PACKAGE_VERSION=$(cat $__scriptpath/BuildToolsVersion.txt)
__DOTNET_TOOLS_VERSION=$(cat $__scriptpath/DotnetCLIVersion.txt)
__PACKAGES_DIR=$__scriptpath/packages
__TOOLRUNTIME_DIR=$__scriptpath/Tools
__DOTNET_PATH=$__TOOLRUNTIME_DIR/dotnetcli
__BUILD_TOOLS_PATH=$__PACKAGES_DIR/Microsoft.DotNet.BuildTools/$__BUILD_TOOLS_PACKAGE_VERSION/lib
__PROJECT_JSON_PATH=$__TOOLRUNTIME_DIR/$__BUILD_TOOLS_PACKAGE_VERSION
__PROJECT_JSON_FILE=$__PROJECT_JSON_PATH/project.csproj

# log file
__init_tools_log=$__scriptpath/init-tools.log

# dotnet command
__DOTNET_CMD=$__DOTNET_PATH/dotnet

# files
declare -a arr=(
'<Project Sdk="Microsoft.NET.Sdk">'

'  <PropertyGroup>'
'    <TargetFramework>netcoreapp1.0</TargetFramework>'
'    <AssemblyName>'$__BUILD_TOOLS_PACKAGE_VERSION'</AssemblyName>'
'    <PackageId>'$__BUILD_TOOLS_PACKAGE_VERSION'</PackageId>'
'  </PropertyGroup>'

'  <ItemGroup>'
'    <PackageReference Include="Microsoft.DotNet.BuildTools" Version="'$__BUILD_TOOLS_PACKAGE_VERSION'" />'
'  </ItemGroup>'

'</Project>'
)

__PROJECT_JSON_CONTENTS=${arr[@]}

if [ -z "$__BUILDTOOLS_SOURCE" ]; then
    __BUILDTOOLS_SOURCE=https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json
fi

# on/off flag
__INIT_TOOLS_DONE_MARKER=$__PROJECT_JSON_PATH/done


# ---------------------------------------------------------------------
# OS determination
# ---------------------------------------------------------------------

get_current_linux_name() {
    # Detect Distro
    if [ "$(cat /etc/*-release | grep -cim1 ubuntu)" -eq 1 ]; then
        if [ "$(cat /etc/*-release | grep -cim1 16.04)" -eq 1 ]; then
            echo "ubuntu.16.04"
            return 0
        fi
        if [ "$(cat /etc/*-release | grep -cim1 16.10)" -eq 1 ]; then
            echo "ubuntu.16.10"
            return 0
        fi

        echo "ubuntu"
        return 0
    elif [ "$(cat /etc/*-release | grep -cim1 centos)" -eq 1 ]; then
        echo "centos"
        return 0
    elif [ "$(cat /etc/*-release | grep -cim1 rhel)" -eq 1 ]; then
        echo "rhel"
        return 0
    elif [ "$(cat /etc/*-release | grep -cim1 debian)" -eq 1 ]; then
        echo "debian"
        return 0
    elif [ "$(cat /etc/*-release | grep -cim1 fedora)" -eq 1 ]; then
        if [ "$(cat /etc/*-release | grep -cim1 23)" -eq 1 ]; then
            echo "fedora.23"
            return 0
        fi
        if [ "$(cat /etc/*-release | grep -cim1 24)" -eq 1 ]; then
            echo "fedora.24"
            return 0
        fi
    elif [ "$(cat /etc/*-release | grep -cim1 opensuse)" -eq 1 ]; then
        if [ "$(cat /etc/*-release | grep -cim1 13.2)" -eq 1 ]; then
            echo "opensuse.13.2"
            return 0
        fi
        if [ "$(cat /etc/*-release | grep -cim1 42.1)" -eq 1 ]; then
            echo "opensuse.42.1"
            return 0
        fi
    fi

    # Cannot determine Linux distribution, assuming Ubuntu 14.04.
    echo "ubuntu"
    return 0
}


# ---------------------------------------------------------------------
# set dotnet pkg variables
# ---------------------------------------------------------------------

if [ -z "$__DOTNET_PKG" ]; then
OSName=$(uname -s)
    case $OSName in
        Darwin)
            OS=OSX
            __DOTNET_PKG=dotnet-dev-osx-x64
            ulimit -n 2048
            ;;

        Linux)
            __DOTNET_PKG="dotnet-dev-$(get_current_linux_name)-x64"
            OS=Linux
            ;;

        *)
            echo "Unsupported OS '$OSName' detected. Downloading ubuntu-x64 tools."
            OS=Linux
            __DOTNET_PKG=dotnet-dev-ubuntu-x64
            ;;
  esac
fi


# ---------------------------------------------------------------------
# Dealing
# ---------------------------------------------------------------------

if [ ! -e $__INIT_TOOLS_DONE_MARKER ]; then

    if [ -e $__TOOLRUNTIME_DIR ]; then rm -rf -- $__TOOLRUNTIME_DIR; fi

    partingLine ─
    echo ''
    echo "Restoring BuildTools version $__BUILD_TOOLS_PACKAGE_VERSION..."
    echo "Running: $__scriptpath/init-tools.sh" > $__init_tools_log
    echo ''



    # ---------------------------------------------------------------------
    # download and install dotnet cli for osx
    # ---------------------------------------------------------------------

    if [ ! -e $__DOTNET_PATH ]; then



        __DOTNET_LOCATION="https://dotnetcli.blob.core.windows.net/dotnet/Sdk/master/dotnet-sdk-latest-osx-x64.tar.gz"

        partingLine ─
        echo ''
        echo "Installing dotnet cli..."
        echo "Installing '${__DOTNET_LOCATION}' to '$__DOTNET_PATH/dotnet.tar'" >> $__init_tools_log
        echo ''

        which curl > /dev/null 2> /dev/null
        if [ $? -ne 0 ]; then
            mkdir -p "$__DOTNET_PATH"
            wget -q -O $__DOTNET_PATH/dotnet.tar ${__DOTNET_LOCATION}
        else
            curl --retry 10 -sSL --create-dirs -o $__DOTNET_PATH/dotnet.tar ${__DOTNET_LOCATION}
        fi

        cd $__DOTNET_PATH
        tar -xf $__DOTNET_PATH/dotnet.tar

        cd $__scriptpath
    fi


    # ---------------------------------------------------------------------
    # create project.json file
    # ---------------------------------------------------------------------

    if [ ! -d "$__PROJECT_JSON_PATH" ]; then
        mkdir "$__PROJECT_JSON_PATH"
    fi

    echo $__PROJECT_JSON_CONTENTS > "$__PROJECT_JSON_FILE"


    # ---------------------------------------------------------------------
    # dotnet restore for TOOL folder
    # ---------------------------------------------------------------------

    if [ ! -e $__BUILD_TOOLS_PATH ]; then

        partingLine ─
        echo ''
        echo "Restoring BuildTools version $__BUILD_TOOLS_PACKAGE_VERSION..."
        echo "Running: $__DOTNET_CMD restore \"$__PROJECT_JSON_FILE\" --no-cache --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE $__INIT_TOOLS_RESTORE_ARGS" >> $__init_tools_log
        echo ''

        declare -a arr_dotnetRestore=(

            $__DOTNET_CMD
            restore
            $__PROJECT_JSON_FILE
            --no-cache
            --packages
            $__PACKAGES_DIR
            --source
            $__BUILDTOOLS_SOURCE
            $__INIT_TOOLS_RESTORE_ARGS

        )

        ${arr_dotnetRestore[@]} >> $__init_tools_log


        if [ ! -e "$__BUILD_TOOLS_PATH/init-tools.sh" ]; then

            echo ''
            echo "ERROR: Could not restore build tools correctly. See '$__init_tools_log' for more details." 1>&2
            echo "============= $__init_tools_log ==========="
            cat $__init_tools_log
            echo "============= end of $__init_tools_log ==========="
            echo ''

        fi

    fi


    # ---------------------------------------------------------------------
    # dotnet restore for ToolRuntime folder
    # Copy Portable Targets Over to ToolRuntime folder
    # ---------------------------------------------------------------------


    partingLine ─
    echo ''
    echo 'convert file( dotnet migrate ) ; project.json ---> foo.csproj'
    echo ''

    tmp="/tool-runtime/tool-runtime.csproj ; $__DOTNET_CMD migrate $__BUILD_TOOLS_PATH/tool-runtime"
    sed -i -e "s#/tool-runtime/project.json#$tmp#" $__BUILD_TOOLS_PATH/init-tools.sh


    foo='echo $__MSBUILD_CONTENT_JSON > "${__PORTABLETARGETS_PROJECTJSON}"'
    bar="$__DOTNET_CMD migrate $__TOOLRUNTIME_DIR/generated "
    baz="__PORTABLETARGETS_PROJECTJSON=${__TOOLRUNTIME_DIR}/generated/generated.csproj"
    sed -i -e "s#$foo#$foo ; $bar ; $baz#" $__BUILD_TOOLS_PATH/init-tools.sh


    partingLine ─
    echo ''
    echo "Initializing BuildTools..."
    echo "Running: $__BUILD_TOOLS_PATH/init-tools.sh $__scriptpath $__DOTNET_CMD $__TOOLRUNTIME_DIR" >> $__init_tools_log
    echo ''

    $__BUILD_TOOLS_PATH/init-tools.sh $__scriptpath $__DOTNET_CMD $__TOOLRUNTIME_DIR >> $__init_tools_log

    if [ "$?" != "0" ]; then
        echo "ERROR: An error occured when trying to initialize the tools. Please check '$__init_tools_log' for more details."1>&2
        exit 1
    fi

    touch $__INIT_TOOLS_DONE_MARKER
    echo "Done initializing tools."


    # ---------------------------------------------------------------------
    # prepare for bulding after initializing tools
    # convert file ( dotnet migrate ) : project.json ---> foo.csproj
    # ---------------------------------------------------------------------

    partingLine ─
    echo ''
    echo "prepare for BuildTools..."

    # delete --legacy-packages-directory
    foo='<DnuCoreRestoreCommand>$(DnuCoreRestoreCommand) --legacy-packages-directory</DnuCoreRestoreCommand>'
    bar='<DnuCoreRestoreCommand>$(DnuCoreRestoreCommand)</DnuCoreRestoreCommand>'
    sed -i -e "s|$foo|$bar|" ./dir.props

    # dotnet migrate
    dotnet migrate src/.nuget/project.json >> $__init_tools_log
    dotnet migrate src/Framework/project.json >> $__init_tools_log
    dotnet migrate Samples/TaskUsageLogger/project.json >> $__init_tools_log
    sed -i -e 's#<Project Sdk="Microsoft.NET.Sdk">#<Project Sdk="Microsoft.NET.Sdk" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">#' Samples/TaskUsageLogger/TaskUsageLogger.csproj
    dotnet migrate Samples/Dependency/project.json  >> $__init_tools_log
    sed -i -e 's#<Project Sdk="Microsoft.NET.Sdk">#<Project Sdk="Microsoft.NET.Sdk" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">#' Samples/Dependency/Dependency.csproj

    dotnet migrate targets/roslyn/project.json  >> $__init_tools_log
    cp targets/roslyn/roslyn.csproj targets/roslyn/project.json
    cp targets/roslyn/project.json targets/roslyn/project.lock.json

    dotnet migrate targets/runtimeDependencies/project.json  >> $__init_tools_log
    cp targets/runtimeDependencies/runtimeDependencies.csproj targets/runtimeDependencies/project.json
    cp targets/runtimeDependencies/project.json targets/runtimeDependencies/project.lock.json

    echo ''
    echo "Done prepare for BuildTools."
    partingLine ─

else
    echo "Tools are already initialized"
fi
```
result
```
msbuild [xplat-master]$ make
./cibuild.sh --scope Compile --host Mono --target Mono --config Release
debug: which mono: /Users/callmekohei/mono64/bin/mono
MONO_BIN_DIR: 
** Using mono from /Users/callmekohei/mono64/bin/mono
Mono JIT compiler version 5.11.0 (master/0d9c3dff 2018年 1月16日 火曜日 14時04分58秒 JST)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          supported, not enabled.
	GC:            sgen (concurrent by default)
** Downloading MSBUILD from https://github.com/Microsoft/msbuild/releases/download/mono-hosted-msbuild-v0.03/mono_msbuild_d25dd923839404bd64cc63f420e75acf96fc75c4.zip
──────────────────────────────────────────────────────────────────────────────────────────

Restoring BuildTools version 1.0.27-prerelease-00927-05...

──────────────────────────────────────────────────────────────────────────────────────────

Installing dotnet cli...

──────────────────────────────────────────────────────────────────────────────────────────

Restoring BuildTools version 1.0.27-prerelease-00927-05...

──────────────────────────────────────────────────────────────────────────────────────────

convert file( dotnet migrate ) ; project.json ---> foo.csproj

──────────────────────────────────────────────────────────────────────────────────────────

Initializing BuildTools...

Done initializing tools.
──────────────────────────────────────────────────────────────────────────────────────────

prepare for BuildTools...

Done prepare for BuildTools.
──────────────────────────────────────────────────────────────────────────────────────────

** Rebuilding MSBuild with downloaded binaries

** using MSBuild in path: /Users/callmekohei/tmp8/msbuild/packages/msbuild/MSBuild.exe
** using runtime host in path: /Users/callmekohei/mono64/bin/mono
** /Users/callmekohei/mono64/bin/mono --debug /Users/callmekohei/tmp8/msbuild/packages/msbuild/MSBuild.exe /t:Rebuild "/Users/callmekohei/tmp8/msbuild/build.proj" /p:OS=OSX /p:Configuration=Release-MONO /p:OverrideToolHost=/Users/callmekohei/mono64/bin/mono /verbosity:minimal  /p:CscToolExe=csc.exe /p:CscToolPath=/Users/callmekohei/tmp8/msbuild/packages/msbuild/ /p:DebugType=portable /p:SourceLinkCreate=false /p:DebugType=portable /fl "/flp:v=diag;logfile=/Users/callmekohei/tmp8/msbuild/msbuild_bootstrap_build-Mono.log"
Microsoft (R) Build Engine version 14.1.0.0
Copyright (C) Microsoft Corporation. All rights reserved.

  Building Project(s): build/dirs.proj : Configuration=Release-MONO, Platform=AnyCPU 
  Building Project(s): src/dirs.proj : Configuration=Release-MONO, Platform=AnyCPU 
  Building Project(s): Samples/dirs.proj : Configuration=Release-MONO, Platform=AnyCPU 
  Building Project(s): mono/tasks/build.csproj : Configuration=Release-MONO, Platform=AnyCPU 
  Microsoft.Build.Framework -> /Users/callmekohei/tmp8/msbuild/bin/Release-MONO/AnyCPU/OSX/Output/Microsoft.Build.Framework.dll
/Users/callmekohei/tmp8/msbuild/targets/DeployDependencies.proj(133,5): error : Unexpected character encountered while parsing value: <. Path '', line 0, position 0.
/Users/callmekohei/tmp8/msbuild/targets/DeployDependencies.proj(133,5): error :   at Newtonsoft.Json.JsonTextReader.ParseValue () [0x002b3] in <5179dd1b79f047d0ac657ed8a725efc1>:0 
/Users/callmekohei/tmp8/msbuild/targets/DeployDependencies.proj(133,5): error :   at Newtonsoft.Json.JsonTextReader.Read () [0x0004c] in <5179dd1b79f047d0ac657ed8a725efc1>:0 
/Users/callmekohei/tmp8/msbuild/targets/DeployDependencies.proj(133,5): error :   at Newtonsoft.Json.Linq.JObject.Load (Newtonsoft.Json.JsonReader reader, Newtonsoft.Json.Linq.JsonLoadSettings settings) [0x00013] in <5179dd1b79f047d0ac657ed8a725efc1>:0 
/Users/callmekohei/tmp8/msbuild/targets/DeployDependencies.proj(133,5): error :   at Newtonsoft.Json.Linq.JObject.Load (Newtonsoft.Json.JsonReader reader) [0x00000] in <5179dd1b79f047d0ac657ed8a725efc1>:0 
/Users/callmekohei/tmp8/msbuild/targets/DeployDependencies.proj(133,5): error :   at Microsoft.DotNet.Build.Tasks.PrereleaseResolveNuGetPackageAssets.ExecuteCore () [0x00047] in <9cf12606b0c64908bc25e25593c01175>:0 
/Users/callmekohei/tmp8/msbuild/targets/DeployDependencies.proj(133,5): error :   at Microsoft.DotNet.Build.Tasks.PrereleaseResolveNuGetPackageAssets.Execute () [0x00000] in <9cf12606b0c64908bc25e25593c01175>:0 
CSC : error CS0006: Metadata file '/Users/callmekohei/tmp8/msbuild/bin/Release-MONO/AnyCPU/OSX/Output/Microsoft.Build.Tasks.Core.dll' could not be found [/Users/callmekohei/tmp8/msbuild/mono/facades/build.proj]
CSC : error CS0006: Metadata file '/Users/callmekohei/tmp8/msbuild/packages/Microsoft.Net.Compilers/2.3.1/tools/Microsoft.Build.Tasks.CodeAnalysis.dll' could not be found [/Users/callmekohei/tmp8/msbuild/mono/facades/build.proj]
CSC : error CS0006: Metadata file '/Users/callmekohei/tmp8/msbuild/bin/Release-MONO/AnyCPU/OSX/Output/Microsoft.Build.Utilities.Core.dll' could not be found [/Users/callmekohei/tmp8/msbuild/mono/facades/build.proj]
/Users/callmekohei/tmp8/msbuild/Samples/TaskUsageLogger/TaskUsageLogger.csproj(19,59): error MSB4066: The attribute "Version" in element <PackageReference> is unrecognized.
/Users/callmekohei/tmp8/msbuild/Samples/Dependency/Dependency.csproj(24,63): error MSB4066: The attribute "Version" in element <PackageReference> is unrecognized.
/Users/callmekohei/tmp8/msbuild/Samples/Dependency/Dependency.csproj(24,63): error MSB4066: The attribute "Version" in element <PackageReference> is unrecognized.
  build -> /Users/callmekohei/tmp8/msbuild/bin/Release-MONO/AnyCPU/OSX/Output/Mono.Build.Tasks.dll
make: *** [all-mono] Error 1
```
### problem point

`DeployRuntim` of `DeployDependencies.proj` can not read `csproj`

```xml
  <Target Name="DeployRuntime"
          DependsOnTargets="RestoreRuntimePackages"
          Outputs="%(RuntimeProjectJson.LockFile).IntentionallyDoesNotExistSoThisAlwaysRuns">
    <PrereleaseResolveNuGetPackageAssets AllowFallbackOnTargetSelection="true"
                                         IncludeFrameworkReferences="false"
                                         NuGetPackagesDirectory="$(PackagesDir)"
                                         RuntimeIdentifier="$(NuGetRuntimeIdentifier)"
                                         ProjectLanguage="C#"
                                         ProjectLockFile="@(RuntimeProjectJson->'%(LockFile)')"
                                         TargetMonikers="$(NuGetTargetMoniker)">
      <Output TaskParameter="ResolvedCopyLocalItems" ItemName="ResolvedRuntimeFiles" />
    </PrereleaseResolveNuGetPackageAssets>
```

**Comment by radical** [comment]
`dotnet migrate` is not going to work here, IIUC, as msbuild is still using `dotnet 1.0.0-preview3-003686` (`$repo/DotnetCLIVersion.txt`). [This](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-migrate) suggests that you can specify a preview version also for the the migration - `dotnet migrate -s -v 1.0.0-preview4` (fix up the version). But I don't know whether this would help or not.

Otherwise, I suggest that you wait for https://github.com/Microsoft/msbuild/pull/2868/ to be merged. Maybe you can help test it too.

**Comment by callmekohei** [comment]
@radical 

Thank you!  I just do it!

> I suggest that you wait for Microsoft/msbuild#2868 to be merged.

Yes, yes, yes. Surely. (^_^)/

**Comment by dsyme** [comment]
Fixed by the latest integration
